### DISCUSSION TITLE
Extending the Keycloak Admin Console with links to add-on applications

### DISCUSSION THREAD

**Discussion by jbman** [keycloak/keycloak#27738] [Ideas]
The Account console is currently missing a way to extend it e.g. for configuration of Keycloak extensions or for additional features related to Keycloak entities.

In other discussions like #13598 it is requested to enhance the console with custom tabs. This idea is about a more lightweight integration of custom UIs, so that an admin can just navigate to an add-on application. Context data (e.g. Keyclaok realm) is provided to the add-on application with URL parameters.

Each navigation link opens an add-on web-application which runs alongside the Keycloak instance. It can be implemented with any UI framework or stack of choice. It could also be a fork of the account console. That's how we currently provide some screens admin users.

The add-on application itself needs to authenticate the user against Keycloak and may use the access token to call custom Keycloak extension endpoints.

## Example

At the realm level, add-on UIs would be listed at a tab "Add-ons".
![realm-addons](https://github.com/keycloak/keycloak/assets/789496/584ddf9a-9e73-45d8-9524-adcb77ee07ff)

The admin can navigate to an add-on (e.g. "Create reports"):
<img src="https://github.com/keycloak/keycloak/assets/789496/19092d80-7521-4d38-8281-a781afce4bd0" width="50%"/>

From the add-on it should be possible to go back to the admin console.

## Add-ons at different Keycloak entities

The navigation links could be related to different Keycloak entities, e.g. users, so that an admin could open an add-on which has the Keycloak User ID availble as context. 
Example:
![user-addons](https://github.com/keycloak/keycloak/assets/789496/07d288ce-5440-4fbe-a1d6-5b3c93f531e0)

Add-ons could be supported for realms, users, groups and roles.

## SPI for Add-ons

The SPI implementation should be able to provide 

- the link name 
- the link to the add-on application including custom parameters to pass context date. 

An implementation helper class or default implementation can resolve a localized  link name based on Keycloak realm localization and provide the context with a pre-defined URL parameter (e.g. "realmId=5c40486e-00c7-4837-b9a3-9df727676724").

The SPI implementation could load and pass any Keycloak data via URL parameter (e.g. user ID, user email, firstname and lastname). This is useful if the add-on doesn't load data from Keycloak but just uses the URL parameters (e.g. for sending an email to a user).

The add-on application should provide a link back to the account console, which is passed as additional URL parameter.

## Feedback

What do you think of such an extension mechanism?
Do you have other example use cases where you would use such an integration?
Does it miss anything other, than the direct integration of a screen in the Admin Console?

**Comment by danielFesenmeyer** [comment]
I don't see necessarily see this as an alternative to https://github.com/keycloak/keycloak/discussions/13598, I rather see it as a minimum implementation of the ideas discussed there.
I'd prefer to have the possibility to add tabs anywhere in the UI,  not just on a "Addons Tab". Because I think it's better for UX - but, of course, that depends on the extensions' use case.
Nevertheless, this "separate Addons approach" would definitely improve status quo.

**Comment by jonkoops** [comment]
Note that we're actively working on being able to register custom tabs and other navigational elements in the Admin Console through Java SPIs. @edewit has done some work for this under #23772, which will likely carry over into the Account Console eventually as we unify the customization code.

**Comment by edewit** [comment]
@jbman how about that we introduce another tab type so that you can register a tab anywhere in the console, but instead of configuring the fields you want to have there we put a iframe there and you specify the address?

**Reply by jbman** [reply]
Iframe based integration would also be an option if the extension application is hosted at the same domain (e.g. an own account console theme). However the integrated page then needs to have the look and controls somewhat aligned and also navigating away needs to prevented there. I think that a link with context data is the most basic integration where both applications don't need to much effort and the admin user still can easily find an additional feature (at realm-, user-, group- or role level)
In any case, some context data must be provided to the integrated or linked page (the realm and the group/user/ role if integrated at that level)
